File: app/src/main/java/cn/ifafu/ifafu/base/mvp/BaseFragment.java
package cn.ifafu.ifafu.base.mvp;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

public abstract class BaseFragment<P extends IPresenter> extends Fragment implements IView {

    protected P mPresenter;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(getLayoutId(), container, false);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initData(savedInstanceState);
        if (mPresenter != null) {
            mPresenter.onCreate();
        }
    }

    public abstract int getLayoutId();

    public abstract void initData(Bundle savedInstanceState);

    @Override
    public void showMessage(String msg) {
        Toast.makeText(getContext(), msg, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showMessage(int stringRes) {
        Toast.makeText(getContext(), stringRes, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void openActivity(Intent intent) {
        startActivity(intent);
    }

    @Override
    public void killSelf() {
        if (getActivity() != null) {
            getActivity().finish();
        }
    }
}

File: app/src/main/java/cn/ifafu/ifafu/base/mvvm/BaseFragment.kt
package cn.ifafu.ifafu.base.mvvm

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import cn.ifafu.ifafu.app.Constant
import cn.ifafu.ifafu.ui.login.LoginActivity
import cn.ifafu.ifafu.view.dialog.LoadingDialog
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

abstract class BaseFragment<VDB : ViewDataBinding, VM : BaseViewModel> : Fragment(), UIEvent {

    protected open val loadingDialog by lazy {
        LoadingDialog(requireContext()).apply {
            setText("加载中")
            setCancelable(true)
        }
    }

    protected lateinit var mBinding: VDB
    protected lateinit var mViewModel: VM

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mBinding = DataBindingUtil.inflate(inflater, getLayoutId(), container, false)
        return mBinding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        mBinding.lifecycleOwner = this
        mViewModel = getViewModel()
        mViewModel.event = this
        initFragment(savedInstanceState)
    }

    protected abstract fun getLayoutId(): Int

    protected abstract fun getViewModel(): VM

    protected abstract fun initFragment(savedInstanceState: Bundle?)

    override suspend fun showMessage(message: String) {
        withContext(Dispatchers.Main) {
            Toast.makeText(this@BaseFragment.requireContext(), message, Toast.LENGTH_SHORT).show()
        }
    }

    override suspend fun showMessage(msgId: Int) {
        withContext(Dispatchers.Main) {
            Toast.makeText(this@BaseFragment.requireContext(), msgId, Toast.LENGTH_SHORT).show()
        }
    }

    override suspend fun showDialog() {
        withContext(Dispatchers.Main) {
            loadingDialog.show()
        }
    }

    override suspend fun hideDialog() {
        withContext(Dispatchers.Main) {
            loadingDialog.cancel()
        }
    }

    override suspend fun startLoginActivity() {
        withContext(Dispatchers.Main) {
            startActivityForResult(Intent(this@BaseFragment.requireContext(), LoginActivity::class.java), Constant.ACTIVITY_LOGIN)
            activity?.finish()
        }
    }

    override suspend fun finishIt() {
        withContext(Dispatchers.IO) {
            activity?.finish()
        }
    }
}